packages/foam-workspace-manager/src/utils/utils.ts

```diff
@@ -1,23 +1,103 @@
-// @todo convert this to use ast parsing
-
-// import unified from 'unified';
-// import markdown from 'remark-parse';
-// import wikiLinkPlugin from 'remark-wiki-link';
-// let processor = unified()
-//   .use(markdown, { gfm: true })
-//   .use(wikiLinkPlugin);
-
-import { findTopLevelHeading, findWikilinksInMarkdown } from './markdown-utils';
+import * as path from 'path';
+import unified from 'unified';
+import markdownParse from 'remark-parse';
+import wikiLinkPlugin from 'remark-wiki-link';
+import visit, { CONTINUE, EXIT } from 'unist-util-visit';
+import { Node, Parent } from 'unist';
+import { BubbleLink, Bubble, Foam } from '../core';
 
 // @ts-expect-error
 export function readWorkspaceFile(filename: string): string {
   throw new Error('Not implemented');
 }
 
+// pipeline cache
+let processor: unified.Processor | null = null;
+
+function parse(markdown: string): Node {
+  processor =
+    processor ||
+    unified()
+      .use(markdownParse, { gfm: true })
+      .use(wikiLinkPlugin);
+  return processor.parse(markdown);
+}
+
+export function createBubbleFromMarkdown(uri: string, markdown: string): Bubble {
+  const filename = path.basename(uri);
+  const id = path.parse(filename).name;
+  const tree = parse(markdown)
+  let title = id
+  visit(tree, node => {
```

Comment:
this is also copied from the other PR, I just put it inline because I need the parsing only once